test/test_transformers.py

```diff
@@ -87,37 +87,85 @@ def use_deterministic_algorithims(mode: bool, warn_only: bool):
+            _check_equal(gold, ref, tst, fudge_factor, tensor_name)
+        return
+
+    # Compute error between golden
+    test_error = (golden - test).abs().max()
+    ref_error = (golden - reference).abs().max()
+
+    if torch.isnan(test_error).any() and not torch.isnan(ref_error).any():
+        raise ValueError("Output/Grad with NaN")
+
+    # We take the max between a default tolerance set else where for float32 or the calculated fudge_factor numbers
+    threshold = max(default_atol[torch.float32], ref_error * fudge_factor)
+    if test_error > threshold:
+        name = tensor_name or ""
+        msg = f"{name} Test error {test_error} is greater than threshold {threshold}!"
+        raise ValueError(msg)
+
+
+def check_out_and_grad(
+    out_tuple: Tuple[torch.Tensor, torch.Tensor, torch.Tensor],
+    grad_query_tuple: Tuple[torch.Tensor, torch.Tensor, torch.Tensor],
+    grad_key_tuple: Tuple[torch.Tensor, torch.Tensor, torch.Tensor],
+    grad_value_tuple: Tuple[torch.Tensor, torch.Tensor, torch.Tensor],
+    grad_attn_mask_tuple: Optional[Tuple[torch.Tensor, torch.Tensor, torch.Tensor]] = None,
+    fudge_factors: Optional[Dict[str, float]] = None
+) -> None:
+    """
+    Check output and gradients of attention mechanism tensors.
+    Compares compiled results against reference and low-precision reference tensors.
+
+    :param out_tuple: Tuple of (ref, lp_ref, compiled) for output tensor
+    :param grad_query_tuple: Tuple of (ref, lp_ref, compiled) for query gradient
+    :param grad_key_tuple: Tuple of (ref, lp_ref, compiled) for key gradient
+    :param grad_value_tuple: Tuple of (ref, lp_ref, compiled) for value gradient
+    :param grad_attn_mask_tuple: Optional tuple of (ref, lp_ref, compiled) for attention mask gradient
+    :param fudge_factors: Dictionary of fudge factors for each tensor type (default uses 5.0 for all)
+    """
+    default_fudge_factor = 5.0
+    if fudge_factors is None:
+        fudge_factors = {}
 
-def get_atol(true_value: torch.Tensor, computed_value: torch.Tensor) -> float:
-    deviation = true_value - computed_value
-    atol = torch.abs(deviation).max().item()
-    return atol
+    out_ref, out_lp_ref, out = out_tuple
 
+    with torch.no_grad():
+        _check_equal(out_ref, out_lp_ref, out, fudge_factors.get('out', default_fudge_factor), "Out")
 
-def get_tolerances(
-    true_value: torch.Tensor,
-    computed_value: torch.Tensor,
-    fudge_factor: Optional[float] = None,
-) -> Tuple[float, float]:
-    """Returns the absolute and relative tolerances for comparing two tensors."""
-    fudge_factor = fudge_factor if fudge_factor is not None else 1.0
-    atol = get_atol(true_value, computed_value)
-    rtol = get_rtol(true_value, computed_value)
+        grad_checks = [
+            (grad_query_tuple, "Grad_Query"),
+            (grad_key_tuple, "Grad_Key"),
+            (grad_value_tuple, "Grad_Value")
+        ]
```

Comment:
tiny nit: lol why upper-snake-case here? I think you could avoid the lower() call below if you used lower-snake-case